src/cargo/ops/registry/publish.rs

```diff
@@ -797,28 +797,51 @@ fn package_list(pkgs: impl IntoIterator<Item = PackageId>, final_sep: &str) -> S
 
 fn resolve_registry_or_index(
     opts: &PublishOpts<'_>,
-    just_pkgs: &[&Package],
+    pkgs: &[&Package],
 ) -> CargoResult<Option<RegistryOrIndex>> {
-    Ok(match opts.reg_or_index.clone() {
-        Some(r) => {
-            validate_registry(&just_pkgs, Some(&r))?;
-            Some(r)
-        }
-        None => {
-            let reg = super::infer_registry(&just_pkgs)?;
-            validate_registry(&just_pkgs, reg.as_ref())?;
-            if let Some(RegistryOrIndex::Registry(registry)) = &reg {
-                if registry != CRATES_IO_REGISTRY {
-                    // Don't warn for crates.io.
-                    opts.gctx.shell().note(&format!(
-                        "found `{}` as only allowed registry. Publishing to it automatically.",
-                        registry
-                    ))?;
+    let opt_index_or_registry = opts.reg_or_index.clone();
+    let registry_is_specified_by_any_package = pkgs
+        .iter()
+        .any(|pkg| pkg.publish().as_ref().map(|v| v.len()).unwrap_or(0) > 0);
+
+    let res = match (opt_index_or_registry, registry_is_specified_by_any_package) {
+        // both `package.publish` and `--index` or `--registry` are specified
+        (Some(opt_reg_or_index), true) => {
+            validate_registry(pkgs, Some(&opt_reg_or_index))?;
+
+            match opt_reg_or_index {
+                r @ RegistryOrIndex::Registry(_) => Some(r),
+                r @ RegistryOrIndex::Index(_) => {
+                    opts.gctx.shell().warn(r#"`--index` will ignore registries set by `package.publish` in Cargo.toml, and may cause unexpected push to prohibited registry
+help: use `--registry` instead or set `publish = true` in Cargo.toml to suppress this warning"#)?;
+
+                    Some(r)
                 }
             }
-            reg
         }
-    })
+        // only `package.publish` section
+        (None, true) => {
+            let registry_or_index = super::infer_registry(pkgs)?;
+            if let Some(RegistryOrIndex::Registry(ref registry)) = registry_or_index
+                && registry != CRATES_IO_REGISTRY
+            {
+                // Don't warn for crates.io.
+                opts.gctx.shell().note(&format!(
+                    "found `{}` as only allowed registry. Publishing to it automatically.",
+                    registry
+                ))?;
+            }
+
+            registry_or_index
+        }
+        (Some(opt_reg), false) => {
+            validate_registry(pkgs, Some(&opt_reg))?;
+            Some(opt_reg)
+        }
+        (None, false) => None,
```

Comment:
The divergence of the `None` cases, particularly lack of `validate_registry` seems unrelated to this change.

Reply:
This still ended up changing the `None` branch, requiring inspection to see that it seems to just be refactors.  That is unrelated to the change at hand and really should be split out to a separate commit.

Please try to separate refactors from the actual behavior change.